# Degree of Freedom

It tells that, how many variables can do freely independently

In [5]:
#Using the Python code
n1 = 10
n2 = 15

degree_of_freedom = n1 + n2 - 2
print(f"Degree of freedom: {degree_of_freedom}")

Degree of freedom: 23


In [6]:
#Using the scipy 

In [7]:
from scipy.stats import ttest_ind
group1 = [10, 12, 14, 15, 13, 11, 16, 14, 12, 15]
group2 = [18, 20, 17, 19, 21, 22, 18, 20, 19, 21, 23, 18, 20, 22, 19]


In [8]:
result = ttest_ind(group1, group2, equal_var=True)

print("t-statistic:", result.statistic)
print("p-value:", result.pvalue)

t-statistic: -8.893581363752412
p-value: 6.654885029375391e-09


Here, because there are 10 + 15 observations, the degrees of freedom are:

d
f
=
10
+
15
−
2
=
23

If you use equal_var=False (Welch's t-test), the degrees of freedom are calculated differently and may not be an integer.

# Chi-squared Test and Cramer's V
$$
\chi^2 = \sum \frac{(O-E)^2}{E}
$$


Where,

- O => Observed,
- E => Expected



For Cramer's V

$$
V = \sqrt{\frac{\chi^2}{n(k-1)}}
$$

To find the relationship between two categorical variables, but it will not show how strong the variables. For that,
we use Cramer's v, it tell is there any association between them.


What the Boundaries Mean
• 0.0: Indicates no association at all; the variables are completely independent.
• 1.0: Indicates a perfect association; knowing the category of one variable allows you to predict the other with 100% certainty

In [9]:
import numpy as np
from scipy.stats import chi2_contingency

In [10]:
# Rows = Gender
# Columns = Preference
data = np.array([
    [30, 20],  # Male
    [10, 40]   # Female
])

In [16]:
chi2, p, df, expected = chi2_contingency(data)

#Also have the sample
k = data.sum()

#Using the formula of the cramers
cramer_v = np.sqrt(chi2/(n*(k - 1)))

In [17]:
print(f"Chi-Squared: {chi2}")
print(f"P-value: {p}")
print(f"Degree of freedom: {df}")
print(f"Expected Frequency: {expected}")
print("======================================")
print(f"Cramers'V is: {cramer_v}")

Chi-Squared: 15.041666666666668
P-value: 0.00010516355403363114
Degree of freedom: 1
Expected Frequency: [[20. 30.]
 [20. 30.]]
Cramers'V is: 0.038978972143455366


In [13]:
if p < 0.05:
    print("There is a significant relationship between the variables.")
else:
    print("There is no significant relationship between the variables.")

There is a significant relationship between the variables.


Chi-Squared: 15.041666666666668, shows that indicates a highly statistically significant result under most standard scenarios, meaning your observed data deviates strongly from what would be expected by chance.

P-value: Reject the null hypothesis; move to the alternative. Hence, there is a significant relationship between the variables.

Degree of freedom: How many variables can vary freely and independently.

Cramers'V is: 0.038978972143455366, there is weak association between two categorical variables.